test/explorer/series/struct_test.exs

```diff
@@ -0,0 +1,147 @@
+    end
+
+    test "errors when structs have mismatched types" do
+      assert_raise ArgumentError,
+                   "the value %{a: \"a\"} does not match the inferred series dtype {:struct, %{\"a\" => :integer}}",
+                   fn -> Series.from_list([%{a: 1}, %{a: "a"}]) end
+
+      assert_raise ArgumentError,
+                   "the value %{b: 1} does not match the inferred series dtype {:struct, %{\"a\" => :integer}}",
+                   fn -> Series.from_list([%{a: 1}, %{b: 1}]) end
+    end
+  end
+
+  describe "cast/2" do
+    test "struct with integers to struct with floats" do
+      s = Series.from_list([%{a: 1}, %{a: 2}])
+      s1 = Series.cast(s, {:struct, %{"a" => {:f, 64}}})
+
+      assert Series.to_list(s1) == [%{"a" => 1.0}, %{"a" => 2.0}]
+      assert Series.dtype(s1) == {:struct, %{"a" => {:f, 64}}}
+    end
+
+    test "nested structs with integers to nested structs with floats" do
+      s = Series.from_list([%{a: %{b: 1}}, %{a: %{b: 2}}])
+      s1 = Series.cast(s, {:struct, %{"a" => {:struct, %{"b" => {:f, 64}}}}})
+
+      assert Series.to_list(s1) == [%{"a" => %{"b" => 1.0}}, %{"a" => %{"b" => 2.0}}]
+      assert Series.dtype(s1) == {:struct, %{"a" => {:struct, %{"b" => {:f, 64}}}}}
+    end
+
+    test "structs with integers to structs with datetimes" do
+      s =
+        Series.from_list([
+          %{a: 1},
+          %{a: 2},
+          %{a: 3},
+          %{a: 1_649_883_642 * 1_000 * 1_000}
+        ])
+
+      s1 = Series.cast(s, {:struct, %{"a" => {:datetime, :microsecond}}})
+
+      assert Series.to_list(s1) == [
+               %{"a" => ~N[1970-01-01 00:00:00.000001]},
+               %{"a" => ~N[1970-01-01 00:00:00.000002]},
+               %{"a" => ~N[1970-01-01 00:00:00.000003]},
+               %{"a" => ~N[2022-04-13 21:00:42.000000]}
+             ]
+
+      assert Series.dtype(s1) == {:struct, %{"a" => {:datetime, :microsecond}}}
+    end
+
+    test "errors when casting to invalid nested types" do
+      s = Series.from_list([%{a: 1}, %{a: 2}])
+
+      assert_raise ArgumentError,
+                   ~r"Explorer.Series.cast/2 not implemented for dtype {:struct, %{\"a\" => :invalid_type}}",
+                   fn -> Series.cast(s, {:struct, %{"a" => :invalid_type}}) end
+    end
+  end
+
+  describe "inspect/1" do
+    test "struct with integer values" do
+      s = Series.from_list([%{a: 1}, %{a: 2}])
+
+      assert inspect(s) ==
+               """
+               #Explorer.Series<
+                 Polars[2]
+                 struct[1] [%{"a" => 1}, %{"a" => 2}]
+               >\
+               """
+    end
+  end
+
+  test "struct with nested values" do
+    s = Series.from_list([%{a: %{b: 1}, c: [2]}, %{a: %{b: 2}, c: [4]}])
+
+    assert inspect(s) ==
+             """
+             #Explorer.Series<
+               Polars[2]
+               struct[2] [%{"a" => %{"b" => 1}, "c" => [2]}, %{"a" => %{"b" => 2}, "c" => [4]}]
+             >\
+             """
+  end
```

Comment:
I was thinking about a crazy scenario where people try something like `{:list, {:struct, any()}` :sweat_smile: 
Would this work? If so, would you mind to add a test case for that?

Reply:
When you say `{:list, {:struct, any()}}` do you mean mixing datatypes? Something like:

```elixir
Series.from_list([[%{a: 1}, %{a: "b"}]])
# or
Series.from_list([[%{a: 1}], [%{a: "b"}]])
```

If so, this should give a type error when building the series. I can add a test to make sure that's what happens, just in case 😅

Reply:
The idea would be to have a nested dtype mixing struct and list. It should be valid:

```elixir
Series.from_list([[%{a: 1}, %{a: 2}]])
# or
Series.from_list([[%{a: 1}], [%{a: 2}]])
```

This should give a `{:list, {:struct, %{"a" => :integer}}}` dtype, right?

Reply:
Oh yeah, this is valid and should work. I can add a test for it just in case (I think I only have the "list inside of a struct" case covered).

Reply:
Done here: [a5eb947](https://github.com/elixir-explorer/explorer/pull/756/commits/a5eb947fcaac4a9d09e6607bec76829a652275a2)